File: app/channels/online_channel.rb
class OnlineChannel < ApplicationCable::Channel
  def subscribed
    # Save online status
    ActionCable.server.pubsub.redis_connection_for_subscriptions.sadd "online", current_user.id

    stream_from "online:users"

    html = ApplicationController.render(partial: "users/online", locals: { user: current_user })
    broadcast_to "users", { id: current_user.id, status: "online", html: html }
  end

  def unsubscribed
    # Remove online status
    ActionCable.server.pubsub.redis_connection_for_subscriptions.srem "online", current_user.id

    broadcast_to "users", { id: current_user.id, status: "offline" }
  end
end

File: app/javascript/channels/online_channel.js
import consumer from "./consumer"

consumer.subscriptions.create("OnlineChannel", {
  connected() {
    // Called when the subscription is ready for use on the server
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    // Called when there's incoming data on the websocket for this channel
    console.log(data)

    let online = document.querySelector("#online-users")
    let el = online.querySelector(`[data-id='${data.id}']`)

    if (data.status == "online" && el == null) {
      online.insertAdjacentHTML('beforeend', data.html)

    } else if (data.status == "offline" && el != null) {
      el.remove()
    }
  }
});
